src/Module.Server/Modes/DTV/CrpgDTVServer.cs

```diff
@@ -0,0 +1,333 @@
+    public override bool IsGameModeUsingOpposingTeams => true;
+    public override bool AllowCustomPlayerBanners() => false;
+    public override bool UseRoundController() => true;
+
+    public CrpgDTVServer(CrpgDTVClient dtvClient,
+        CrpgRewardServer rewardServer,
+        CrpgDTVTeamSelectComponent teamSelectComponent,
+        CrpgDTVSpawningBehavior spawningBehavior)
+    {
+        _dtvClient = dtvClient;
+        _rewardServer = rewardServer;
+        _teamSelectComponent = teamSelectComponent;
+        _spawningBehavior = spawningBehavior;
+    }
+
+    public override MissionLobbyComponent.MultiplayerGameType GetMissionType()
+    {
+        return MissionLobbyComponent.MultiplayerGameType.Battle;
+    }
+
+    public override void AfterStart()
+    {
+        base.AfterStart();
+        RoundController.OnPreRoundEnding += OnPreRoundEnding;
+        AddTeams();
+    }
+
+    public override void OnBehaviorInitialize()
+    {
+        base.OnBehaviorInitialize();
+        // TODO: SetTeamColorsWithAllSynched
+    }
+
+    public override void OnRemoveBehavior()
+    {
+        RoundController.OnPreRoundEnding -= OnPreRoundEnding;
+        base.OnRemoveBehavior();
+    }
+
+    public override void OnClearScene()
+    {
+    }
+
+    public override bool CheckForWarmupEnd()
+    {
+        int playersInTeam = 0;
+        foreach (NetworkCommunicator networkPeer in GameNetwork.NetworkPeers)
+        {
+            MissionPeer component = networkPeer.GetComponent<MissionPeer>();
+            if (networkPeer.IsSynchronized && component?.Team != null && component.Team.Side != BattleSideEnum.None)
+            {
+                playersInTeam += 1;
+            }
+        }
+
+        return playersInTeam >= MultiplayerOptions.OptionType.MaxNumberOfPlayers.GetIntValue();
+    }
+
+    public override void OnMissionTick(float dt)
+    {
+        base.OnMissionTick(dt);
+        if (MissionLobbyComponent.CurrentMultiplayerState != MissionLobbyComponent.MultiplayerGameState.Playing
+            || !RoundController.IsRoundInProgress
+            || !CanGameModeSystemsTickThisFrame)
+        {
+            return;
+        }
+
+        CheckForAttackers();
+
+        if (BotRespawnDelayEnded())
+        {
+            SpawnWave(currentRound, currentWave);
+            _waitingForBotSpawn = false;
+        }
+    }
+
+    public void CheckForAttackers()
+    {
+        bool attackersDepleted = !Mission.AttackerTeam.HasBots;
+
+        if (attackersDepleted && !_waitingForBotSpawn)
+        {
+            Debug.Print("Attackers depleted");
+            currentWave += 1;
+            UpdateMessages();
+            if (currentWave >= TotalWaves + 1)
+            {
+                OnRoundEnd();
+            }
+            else
+            {
+                OnWaveEnd();
+            }
+
+            if (currentRound > TotalRounds)
+            {
+                Debug.Print("Match complete");
+                // end match
```

Comment:
Is that a TODO?

Reply:
removed as the match complete check is caught in OnPreRoundEnding()